### ISSUE TITLE
Reading HTTP Response When Request Isn't Fully Sent

### ISSUE THREAD

**Issue by dzelemba** [square/okhttp#4425]
If I send a large POST request using OkHttp, but the server responds quickly with an error (e.g. permission denied) I get a

java.net.SocketException: Broken pipe (Write failed)( status = 272 )

exception.

Command line tools like curl seem to handle this case just fine and are able to return the HTTP Response instead of aborting because the write failed. Can okhttp do the same?

Note: also asked here https://stackoverflow.com/questions/53451467/reading-http-response-when-request-isnt-fully-sent.

**Comment by swankjesse** [comment]
Tough one. I’m not sure.

**Comment by bstsnail** [comment]
I have the same problem, still don't know how to get the response

**Comment by yschimke** [comment]
Won't fix.

**Comment by drtrang** [comment]
> Won't fix.

Why? It confuses me.

**Comment by dzelemba** [comment]
For anyone struggling with this issue, we did end up getting it to work using Jetty and HTTP 2.0: https://github.com/GoogleCloudPlatform/healthcare-dicom-dicomweb-adapter/blob/master/dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java#L65

**Comment by swankjesse** [comment]
Related: https://github.com/square/okhttp/pull/6121

**Comment by jan-berge-ommedal** [comment]
@swankjesse I have been testing the latest alpha-release and was happy to see early server responses handled much better than before! The fix works great when the server decides to end the exchange early.

However, i have another use-case where the request body source stream might fail client-side during transmission. In this case, an exception is thrown because of:

https://github.com/square/okhttp/blob/master/okhttp/src/main/kotlin/okhttp3/internal/http/CallServerInterceptor.kt#L86
```
	if (!exchange.hasFailure) {
        throw e // Don't attempt to read the response; we failed to send the request.
	}
```
even if we actually did manage to partially send the request.

It would be nice to still be able to get the server response code (typically 4xx/5xx) + headers/trailers in this case. I started sketching on a way to fix this: https://github.com/square/okhttp/compare/master...jan-berge-ommedal:partial_requests

Could something like this be fitted into okhttp?
